File: src/linked_list.rb
require 'ffi'

module LinkedList
  extend FFI::Library
  ffi_lib "linked_list/linked_list.so"
 
  class Element < FFI::Struct
    layout :next, :pointer,
           :element, :int
  end
  
  class LinkedList < FFI::Struct
    layout :front, :pointer
  end

  attach_function :ll_element_init, [:int], Element.by_ref
  attach_function :ll_element_destroy, [Element.by_ref], :void

  attach_function :linked_list_init, [], LinkedList.by_ref
  attach_function :linked_list_destroy, [LinkedList.by_ref], :void

  attach_function :insert_key, [:int,LinkedList.by_ref], :void
  attach_function :insert_node, [Element.by_ref,LinkedList.by_ref], :void
  attach_function :has_cycled, [LinkedList.by_ref], Element.by_ref

end

class TestLinkedList
  def empty_list
    list = LinkedList.linked_list_init
    result = LinkedList.has_cycled(list).null?
    LinkedList.linked_list_destroy(list)
    result
  end
  
  def single_node_no_cycle
    list = LinkedList.linked_list_init
    LinkedList.insert_key(rand(100),list)
    result = LinkedList.has_cycled(list).null?
    LinkedList.linked_list_destroy(list)
    result
  end

  def single_node_cycle
    list = LinkedList.linked_list_init
    node = LinkedList.ll_element_init(rand(10000))
    2.times {
      LinkedList.insert_node(node,list)
    }
    result = LinkedList.has_cycled(list).null?
    LinkedList.linked_list_destroy(list)
    !result
  end

  def huge_list_no_cycle
    list = LinkedList.linked_list_init
    
    ( 1000 + rand(1000)).times { LinkedList.insert_key(rand(10000),list) }
    result = LinkedList.has_cycled(list).null?
    LinkedList.linked_list_destroy(list)
    result
  end

  def huge_list_cycle
    list = LinkedList.linked_list_init
    size = rand(1000)
    choice = rand(size)
    element = LinkedList.ll_element_init(rand(1000))
    last_element = LinkedList.ll_element_init(rand(1000))
    size.times {
      LinkedList.insert_node(element, list) if (choice)
      LinkedList.insert_key(rand(1000),list)
    }
    LinkedList.insert_node(last_element,list)
    last_element[:next] = last_element.pointer

    result = LinkedList.has_cycled(list).null?
    LinkedList.linked_list_destroy(list)
    !result
  end
  
end

File: src/linked_list/linked_list.c
#include "linked_list.h"

#include <stdlib.h>

struct ll_element* ll_element_init(int element) 
{
	struct ll_element* node = 
		(struct ll_element*) malloc(sizeof(struct ll_element));

	node->element = element;
	node->next = NULL;
	
	return node;
}

void ll_element_destroy(struct ll_element* node)
{
	free(node);
}

struct linked_list* linked_list_init() 
{

	struct linked_list* list = 
		(struct linked_list*) malloc(sizeof(struct linked_list));
	
	list->front = NULL;

	return list;
}

void linked_list_destroy(struct linked_list* list) 
{
	struct ll_element *tmp1, *tmp2, *begin;
	tmp1 = list->front;
	
	if (has_cycled(list))
		begin = tmp1;

	for (tmp1 = list->front; tmp1 != NULL; tmp1 = tmp2) {
		tmp2 = tmp1->next;
		if (begin != tmp1)
			free(tmp1);
		else
			break;
	}

	free(list);
} 

void insert_key(int element, struct linked_list* list) 
{
	insert_node(ll_element_init(element), list);	
}

void insert_node(struct ll_element* element, struct linked_list* list)
{
	if (element == NULL || list == NULL) 
		return;

	element->next = list->front;
	list->front = element;
}


struct ll_element*  has_cycled(struct linked_list* list)
{	
	struct ll_element *hare, *turtle;

	hare = turtle = list->front;
	int count = FALSE;
	
	while (hare != NULL) {
		if (count) {
			turtle = turtle->next;
			count  = FALSE;
		} else {
			count = TRUE;
		}
		hare = hare->next;
		
		if (hare == turtle) 
			return hare;
	}
	
	return NULL;
}
